src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11800,6 +11869,77 @@ private ExpressionSyntax ParseNewExpression()
             }
         }
 
+        private CollectionCreationExpressionSyntax ParseCollectionCreationExpression()
+        {
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.OpenBracketToken);
+            var openBracket = this.EatToken(SyntaxKind.OpenBracketToken);
+            var list = this.ParseCommaSeparatedSyntaxList(
+                ref openBracket,
+                SyntaxKind.CloseBracketToken,
+                static @this => @this.IsPossibleCollectionElement(),
+                static @this => @this.ParseCollectionElement(),
+                skipBadCollectionElementTokens,
+                allowTrailingSeparator: true,
+                requireOneElement: false,
+                allowSemicolonAsSeparator: false);
+
+            return _syntaxFactory.CollectionCreationExpression(
+                openBracket,
+                list,
+                this.EatToken(SyntaxKind.CloseBracketToken));
+
+            static PostSkipAction skipBadCollectionElementTokens(
+                LanguageParser @this, ref SyntaxToken openBracket, SeparatedSyntaxListBuilder<CollectionElementSyntax> list, SyntaxKind expectedKind, SyntaxKind closeKind)
+            {
+                return @this.SkipBadSeparatedListTokensWithExpectedKind(ref openBracket, list,
+                    static p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleCollectionElement(),
+                    static (p, closeKind) => p.CurrentToken.Kind == closeKind,
```

Comment:
The shadowing of `closeKind` is confusing. Let's use different identifiers.

Reply:
this shadowing is *extremely* intentional (and done with all callback lambdas).  It's precisely to indicate taht it's the same value, just passed through the lambda as an arg so that you always have hte same gy.  Renaming is actually more confusing here as now there's the potential semantic confusion of "oh wait... is this closeKind2 somehow different from the containing closeKind1, and what does that mean?"

As with teh above, if we are to change this, i woulid prefer it be in a separate PR that does that for all the code like this (there are at least a dozen places to fixup if we do this).